fontforge/parsettf.c

```diff
@@ -4241,71 +4241,73 @@ return( -1 );
     if ( modtype<=1 /* Unicode */ ) {
 	/* No conversion needed, already unicode */;
     } else if ( modtype==2 /* SJIS */ ) {
-	if ( enc<=127 ) {
-	    /* Latin */
-	    if ( enc=='\\' ) enc = 0xa5;	/* Yen */
-	} else if ( enc>=161 && enc<=223 ) {
-	    /* Katakana */
-	    enc = unicode_from_jis201[enc];
-	} else if ( enc<255 ) {
-	    /* This is erroneous as I understand SJIS */
-	    enc = badencoding(info);
-	} else if (enc >= 0xeaa5) {
-        /* Encoded value is outside SJIS range */
-        /* If this happens, it's likely that it's actually CP932 encoded */
-        /* Todo: Detect CP932 encoding earlier and apply that instead of SJIS */
-        enc = badencoding(info);
-	} else {
-	    int ch1 = enc>>8, ch2 = enc&0xff;
-	    if ( ch1 >= 129 && ch1<= 159 )
-		ch1 -= 112;
-	    else
-		ch1 -= 176;
-	    ch1 <<= 1;
-	    if ( ch2>=159 )
-		ch2-= 126;
-	    else if ( ch2>127 ) {
-		--ch1;
-		ch2 -= 32;
-	    } else {
-		--ch1;
-		ch2 -= 31;
-	    }
-	    if ( ch1<0x21 || ch2<0x21 || ch1>0x7e || ch2>0x7e )
-		enc = badencoding(info);
-	    else
-		enc = unicode_from_jis208[(ch1-0x21)*94+(ch2-0x21)];
-	}
-    } else if ( modtype==3 /* GB2312 offset by 0x8080, parse just like wansung */ ) {
-	if ( enc>0xa1a1 ) {
-	    enc -= 0xa1a1;
-	    enc = (enc>>8)*94 + (enc&0xff);
-	    enc = unicode_from_gb2312[enc];
-	    if ( enc==0 ) enc = -1;
-	} else if ( enc>0x100 )
-	    enc = badencoding(info);
+        // Apart from these, cp932 is a strict superset of sjis
```

Comment:
This section may be the most contentious part, using  iconv for reading the cmap encoding. But at least for SJIS, I think this is actually better. There are some encoding differences for gb2312/big5/wansung/johab, but it's hard to say if it's for the better or worse.

From testing purely the conversion routines, our custom rolled conversion tables were spitting out garbage conversions for at least some slots, but I don't know if that had any effect in practice. Also I guess iconv support may not be equal everywhere. But I believe we're already using iconv for everything else as-is (particularly for re-encoding).

This was/is really hard to test, non-unicode encoded fonts are so rare today.

Reply:
Using iconv is better because at least font developers know what iconv's choices are and what to look out for. Using our own tables, based just on George's opinions, makes things more confusing. So I agree with this change.